File: web/src/model/victron_controller/dashboard/Owner.ts
import {BaboonCodecContext, BaboonBinWriter, BinTools, BaboonBinReader, Lazy} from '../../BaboonSharedRuntime'

export enum Owner {
    Unset = "Unset",
    System = "System",
    Dashboard = "Dashboard",
    HaMqtt = "HaMqtt",
    WeatherSocPlanner = "WeatherSocPlanner",
    SetpointController = "SetpointController",
    CurrentLimitController = "CurrentLimitController",
    ScheduleController = "ScheduleController",
    ZappiController = "ZappiController",
    EddiController = "EddiController",
    FullChargeScheduler = "FullChargeScheduler"
}

export const Owner_values: ReadonlyArray<Owner> = [
    Owner.Unset,
    Owner.System,
    Owner.Dashboard,
    Owner.HaMqtt,
    Owner.WeatherSocPlanner,
    Owner.SetpointController,
    Owner.CurrentLimitController,
    Owner.ScheduleController,
    Owner.ZappiController,
    Owner.EddiController,
    Owner.FullChargeScheduler
] as const;

export function Owner_parse(s: string): Owner {
    const found = Owner_values.find(v => v === s);
    if (found === undefined) {
        throw new Error("Unknown Owner variant: " + s);
    }
    return found;
}

export class Owner_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: Owner, writer: BaboonBinWriter): unknown {
        if (this !== Owner_UEBACodec.lazyInstance.value) {
          return Owner_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        switch (value) {
            case "Unset": BinTools.writeByte(writer, 0); break;
                case "System": BinTools.writeByte(writer, 1); break;
                case "Dashboard": BinTools.writeByte(writer, 2); break;
                case "HaMqtt": BinTools.writeByte(writer, 3); break;
                case "WeatherSocPlanner": BinTools.writeByte(writer, 4); break;
                case "SetpointController": BinTools.writeByte(writer, 5); break;
                case "CurrentLimitController": BinTools.writeByte(writer, 6); break;
                case "ScheduleController": BinTools.writeByte(writer, 7); break;
                case "ZappiController": BinTools.writeByte(writer, 8); break;
                case "EddiController": BinTools.writeByte(writer, 9); break;
                case "FullChargeScheduler": BinTools.writeByte(writer, 10); break;
            default: throw new Error("Unknown enum variant: " + value);
        }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): Owner {
        if (this !== Owner_UEBACodec .lazyInstance.value) {
            return Owner_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const tag = BinTools.readByte(reader);
        switch (tag) {
            case 0: return "Unset" as Owner;
                case 1: return "System" as Owner;
                case 2: return "Dashboard" as Owner;
                case 3: return "HaMqtt" as Owner;
                case 4: return "WeatherSocPlanner" as Owner;
                case 5: return "SetpointController" as Owner;
                case 6: return "CurrentLimitController" as Owner;
                case 7: return "ScheduleController" as Owner;
                case 8: return "ZappiController" as Owner;
                case 9: return "EddiController" as Owner;
                case 10: return "FullChargeScheduler" as Owner;
            default: throw new Error("Unknown enum variant tag: " + tag);
        }
    }

    public static readonly BaboonDomainVersion = '0.1.0'
    public baboonDomainVersion() {
        return Owner_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return Owner_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#Owner'
    public baboonTypeIdentifier() {
        return Owner_UEBACodec.BaboonTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new Owner_UEBACodec())
    public static get instance(): Owner_UEBACodec {
        return Owner_UEBACodec.lazyInstance.value
    }
}

File: crates/dashboard-model/src/victron_controller/dashboard/owner.rs


#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
pub enum Owner {
    Unset,
    System,
    Dashboard,
    HaMqtt,
    WeatherSocPlanner,
    SetpointController,
    CurrentLimitController,
    ScheduleController,
    ZappiController,
    EddiController,
    FullChargeScheduler,
}

impl Owner {
    pub fn parse(s: &str) -> Result<Self, String> {
        match s {
            "Unset" => Ok(Owner::Unset),
            "System" => Ok(Owner::System),
            "Dashboard" => Ok(Owner::Dashboard),
            "HaMqtt" => Ok(Owner::HaMqtt),
            "WeatherSocPlanner" => Ok(Owner::WeatherSocPlanner),
            "SetpointController" => Ok(Owner::SetpointController),
            "CurrentLimitController" => Ok(Owner::CurrentLimitController),
            "ScheduleController" => Ok(Owner::ScheduleController),
            "ZappiController" => Ok(Owner::ZappiController),
            "EddiController" => Ok(Owner::EddiController),
            "FullChargeScheduler" => Ok(Owner::FullChargeScheduler),
            _ => Err(format!("Unknown variant: {}", s)),
        }
    }

    pub fn all() -> Vec<Owner> {
        vec![
            Owner::Unset,
            Owner::System,
            Owner::Dashboard,
            Owner::HaMqtt,
            Owner::WeatherSocPlanner,
            Owner::SetpointController,
            Owner::CurrentLimitController,
            Owner::ScheduleController,
            Owner::ZappiController,
            Owner::EddiController,
            Owner::FullChargeScheduler,
        ]
    }
}

impl std::fmt::Display for Owner {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Owner::Unset => write!(f, "Unset"),
            Owner::System => write!(f, "System"),
            Owner::Dashboard => write!(f, "Dashboard"),
            Owner::HaMqtt => write!(f, "HaMqtt"),
            Owner::WeatherSocPlanner => write!(f, "WeatherSocPlanner"),
            Owner::SetpointController => write!(f, "SetpointController"),
            Owner::CurrentLimitController => write!(f, "CurrentLimitController"),
            Owner::ScheduleController => write!(f, "ScheduleController"),
            Owner::ZappiController => write!(f, "ZappiController"),
            Owner::EddiController => write!(f, "EddiController"),
            Owner::FullChargeScheduler => write!(f, "FullChargeScheduler"),
        }
    }
}

impl std::str::FromStr for Owner {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::parse(s)
    }
}

impl crate::baboon_runtime::BaboonBinCodecIndexed for Owner {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        0
    }
}

impl crate::baboon_runtime::BaboonBinEncode for Owner {
    fn encode_ueba(&self, _ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        match self {
            Owner::Unset => crate::baboon_runtime::bin_tools::write_byte(writer, 0)?,
            Owner::System => crate::baboon_runtime::bin_tools::write_byte(writer, 1)?,
            Owner::Dashboard => crate::baboon_runtime::bin_tools::write_byte(writer, 2)?,
            Owner::HaMqtt => crate::baboon_runtime::bin_tools::write_byte(writer, 3)?,
            Owner::WeatherSocPlanner => crate::baboon_runtime::bin_tools::write_byte(writer, 4)?,
            Owner::SetpointController => crate::baboon_runtime::bin_tools::write_byte(writer, 5)?,
            Owner::CurrentLimitController => crate::baboon_runtime::bin_tools::write_byte(writer, 6)?,
            Owner::ScheduleController => crate::baboon_runtime::bin_tools::write_byte(writer, 7)?,
            Owner::ZappiController => crate::baboon_runtime::bin_tools::write_byte(writer, 8)?,
            Owner::EddiController => crate::baboon_runtime::bin_tools::write_byte(writer, 9)?,
            Owner::FullChargeScheduler => crate::baboon_runtime::bin_tools::write_byte(writer, 10)?,
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for Owner {
    fn decode_ueba(_ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let tag = crate::baboon_runtime::bin_tools::read_byte(reader)?;
        match tag {
            0 => Ok(Owner::Unset),
            1 => Ok(Owner::System),
            2 => Ok(Owner::Dashboard),
            3 => Ok(Owner::HaMqtt),
            4 => Ok(Owner::WeatherSocPlanner),
            5 => Ok(Owner::SetpointController),
            6 => Ok(Owner::CurrentLimitController),
            7 => Ok(Owner::ScheduleController),
            8 => Ok(Owner::ZappiController),
            9 => Ok(Owner::EddiController),
            10 => Ok(Owner::FullChargeScheduler),
            _ => Err(format!("Unknown enum variant tag: {}", tag).into()),
        }
    }
}